apps/ca.c

```diff
@@ -425,6 +435,20 @@ int ca_main(int argc, char **argv)
         case OPT_ENGINE:
             e = setup_engine(opt_arg(), 0);
             break;
+        case OPT_SM2ID:
+            /* we assume the input is not a hex string */
+            sm2_id = (unsigned char *)opt_arg();
```

Comment:
SHouldnt normally happen, but sm2_free=1 is possible

Reply:
I think the value returned by `opt_arg()` should never be freed, so there is no need to set `sm2_free` to 1.

Reply:
what I meant was if OPT_SM2HEXID allocates data and then stupidly the OPT_SM2ID branch is  called afterwards then sm2_id will swap away from the allocated memory, and then since sm2_free = 1 this will then crash when it tries to free sm2_id.

Reply:
Ahh, that is reasonable. So I think OPT_SM2ID and OPT_SM2HEXID should be mutual exclusive...

Reply:
you could also just goto err in both cases if sm2_id was already set.

Reply:
Right. That is exactly what I plan to do

Reply:
Actually I find the check on duplicated sm2_id settings has been done in `req.c`, so I must miss that part when copy-n-paste this piece of code. Already fixed here, nice catch!